File: lua/robot-simulator/robot-simulator.lua
----------------------------------------
-- Robot simulator:
-- place a robot on a map and move it around
----------------------------------------

-- valid headings
local bearings = { east=0, north=90, west=180, south=270 }
local headings = {}
for k,v in pairs(bearings) do headings[v]=k end

----------------------------------------
-- the Robot class
local Robot = {}
Robot.__index = Robot

-- enable object creation with: r = Robot(...)
setmetatable(Robot, {
    __call = function(cls, ...) return cls:new(...) end,
})

function Robot:new(attributes)
    attributes = attributes or {}

    local robot = {
        x = attributes.x or 0,
        y = attributes.y or 0,
        heading = attributes.heading,
    }
    if not bearings[robot.heading] then robot.heading = 'north' end

    setmetatable(robot, self)
    return robot
end

function Robot:move(instructions)
    for instruction in instructions:gmatch('.') do
        if     instruction == 'A' then self:advance()
        elseif instruction == 'L' then self:turn(1)
        elseif instruction == 'R' then self:turn(-1)
        else   error('Unknown instruction')
        end
    end
end

function Robot:advance()
    local rad = math.rad(bearings[self.heading])
    self.x = self.x + math.cos(rad)
    self.y = self.y + math.sin(rad)
end

function Robot:turn(dir)
    local bearing = (bearings[self.heading] + dir*90) % 360
    self.heading = headings[bearing]
end

return Robot

File: python/robot-simulator/robot_simulator.py
from math import radians, cos, sin
from functools import partialmethod

EAST = 0
NORTH = 90
WEST = 180
SOUTH = 270


class Robot(object):
    def __init__(self, direction=NORTH, x=0, y=0):
        self.direction = direction
        self.coordinates = (x, y)

    def turn(self, dir):
        self.direction = (self.direction + 90 * dir) % 360

    turn_right = partialmethod(turn, dir=-1)
    turn_left  = partialmethod(turn, dir=+1)

    def advance(self):
        rad = radians(self.direction)
        x, y = self.coordinates
        dx, dy = int(cos(rad)), int(sin(rad))
        self.coordinates = (x + dx, y + dy)

    def move(self, script):
        dispatch = {
            'R': self.turn_right,
            'L': self.turn_left,
            'A': self.advance
        }
        for cmd in script:
            if cmd in dispatch:
                dispatch[cmd]()
            else:
                raise ValueError(f'Unknown instruction {cmd}')
